### PR TITLE
Fix plugin shutdown handling

### LINKED ISSUE #41: Logstash cloudwatch plugin blocking pipeline reload
We need to periodically collect few metrics from cloudwatch. 
We are using logstash-input-cloudwatch plugin for this purpose. 
Here is how the setup goes: 
Logstash pipeline has a bunch of logstash agents configured and one of them is this with cloudwatch input plugin. Logstash is configured to auto reload config changes. However, we observed that the pipeline restarts is blocked by this input plugin. We see below error message when we try to auto reload pipeline changes or gracefully stop the service. The only way to stop logstash is to issue a "kill -9". 

Did anyone face this issue?
Could someone help with solutions or any ideas towards overcoming.

Thanks in advance !!

### PR DESCRIPTION
**Pull request by robbavey**
Makes call to Stud.stop! to shutdown the loop when the plugin is stopped

Fixes #41

### CODE REVIEW

**Comment by colinsurprenant** [comment]
@robbavey I would advise against that design which has proven to cause concurrency problems and is hard to reason about. See this PR were I fixed such a problem this week logstash-input-heartbeat#15

Would it be possible to use this pattern instead?

**Comment by colinsurprenant** [comment]
@robbavey great! thanks! 
code LGTM but could you also add the input shutdown spec similar to this also? https://github.com/logstash-plugins/logstash-input-heartbeat/blob/cc94ae29c1457d5f4e51a8c1bce3963ad7aebb07/spec/inputs/heartbeat_spec.rb#L6-L8

**Comment by colinsurprenant** [comment]
@robbavey I am also fine if you want to followup with another PR for the spec, your choice.
LGTM

File: spec/inputs/cloudwatch_spec.rb [hunk 1]
```diff
@@ -51,6 +51,19 @@
...
 
+  describe 'shutdown' do
+    let(:metrics) { double("metrics") }
+    let(:config) { super().merge('namespace' => 'AWS/EC2') }
+
+    before do
+      allow(subject).to receive(:metrics_for).and_return(metrics)
+      allow(metrics).to receive(:count).and_return(1)
+      allow(metrics).to receive(:each).and_return(['DiskWriteBytes'])
+    end
```
Comment (colinsurprenant) [inline spec/inputs/cloudwatch_spec.rb]: 👍

**Comment by colinsurprenant** [comment]
LGTM² thanks for the spec @robbavey

**Comment by robbavey** [comment]
Thanks @colinsurprenant!

**Pull request closed without merging**